framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java

```diff
@@ -251,57 +290,119 @@ public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror
         public AnnotationMirror leastUpperBound(AnnotationMirror a1, AnnotationMirror a2) {
             if (!AnnotationUtils.areSameIgnoringValues(
                     getTopAnnotation(a1), getTopAnnotation(a2))) {
+                // The annotations are in different hierarchies
                 return null;
-            } else if (isSubtype(a1, a2)) {
+            }
+
+            if (isSubtype(a1, a2)) {
                 return a2;
             } else if (isSubtype(a2, a1)) {
                 return a1;
             }
-            // If both are the same type, determine the type and merge:
-            else if (AnnotationUtils.areSameIgnoringValues(a1, a2)) {
-                List<Object> a1Values =
-                        AnnotationUtils.getElementValueArray(a1, "value", Object.class, true);
-                List<Object> a2Values =
-                        AnnotationUtils.getElementValueArray(a2, "value", Object.class, true);
-                HashSet<Object> newValues = new HashSet<Object>(a1Values.size() + a2Values.size());
 
-                newValues.addAll(a1Values);
-                newValues.addAll(a2Values);
+            if (AnnotationUtils.areSameIgnoringValues(a1, a2)) {
+                // If both are the same type, determine the type and merge
+                if (AnnotationUtils.areSameByClass(a1, IntRange.class)) {
+                    // special handling for IntRange
+                    Range range1 = getIntRange(a1);
+                    Range range2 = getIntRange(a2);
+                    return createIntRangeAnnotation(range1.union(range2));
+                } else {
+                    List<Object> a1Values =
+                            AnnotationUtils.getElementValueArray(a1, "value", Object.class, true);
+                    List<Object> a2Values =
+                            AnnotationUtils.getElementValueArray(a2, "value", Object.class, true);
+                    HashSet<Object> newValues =
+                            new HashSet<Object>(a1Values.size() + a2Values.size());
 
-                return createAnnotation(a1.getAnnotationType().toString(), newValues);
+                    newValues.addAll(a1Values);
+                    newValues.addAll(a2Values);
+
+                    return createAnnotation(a1.getAnnotationType().toString(), newValues);
```

Comment:
Yes, it is handled somewhere else (ValueAnnotatedTypeFactory#replaceWithNewAnnoInSpecialCases). Will do the comment. Thanks.